Bot.Core/TownMaintenance.cs

```diff
@@ -0,0 +1,106 @@
+        private readonly IDateTime m_dateTime;
+        private readonly IShutdownPreventionService m_shutdownPrevention;
+
+        private readonly TaskCompletionSource m_shutdownTcs = new();
+        private bool m_shutdownRequested = false;
+        private bool m_queueProcessing = false;
+
+        List<Func<TownKey, Task>> m_startupTasks = new();
+
+        public TownMaintenance(IServiceProvider sp)
+        {
+            sp.Inject(out m_dateTime);
+            sp.Inject(out m_townDatabase);
+            sp.Inject(out m_botClient);
+            sp.Inject(out m_shutdownPrevention);
+
+            m_shutdownPrevention.ShutdownRequested += (s, e) =>
+            {
+                m_shutdownRequested = true;
+                if (!m_queueProcessing)
+                {
+                    m_shutdownTcs.TrySetResult();
+                }
+            };
+            m_shutdownPrevention.RegisterShutdownPreventer(m_shutdownTcs.Task);
+
+            var callbackFactory = sp.GetService<ICallbackSchedulerFactory>();
+            m_callbackScheduler = callbackFactory.CreateScheduler<Queue<TownKey>>(ProcessQueue, TimeSpan.FromMinutes(1));
+
+            m_botClient.Connected += async (o, args) => await RunMaintenanceAsync();
+        }
+
+        private async Task ProcessQueue(Queue<TownKey> towns)
+        {
+            m_queueProcessing = true;
+            Serilog.Log.Information("TownMaintenance: {townCount} towns remain...", towns.Count());
+
+            int numSent = 0;
+            while (numSent < NUM_TOWNS_PER_CALLBACK && towns.Count > 0)
+            {
+                var townKey = towns.Dequeue();
+
+                Serilog.Log.Information("TownMaintenance: Processing town {townKey}", townKey);
+
+                foreach (var task in m_startupTasks)
+                {
+                    try
+                    {
+                        await task(townKey);
+                    }
+                    catch (Exception)
+                    {
+                        // todo: something?
+                    }
+                }
+                numSent++;
+            }
+
+            if (towns.Count > 0 && !m_shutdownRequested)
+            {
+                Serilog.Log.Information("TownMaintenance: {townCount} towns remain, scheduling callback...", towns.Count);
+                DateTime nextTime = m_dateTime.Now + TimeSpan.FromMinutes(MINUTES_PER_CALLBACK);
+                m_callbackScheduler.ScheduleCallback(towns, nextTime);
+            }
+            else
+            {
+                Serilog.Log.Information("TownMaintenance: maintenance complete!");
+            }
+
+            if(m_shutdownRequested)
+                m_shutdownTcs.TrySetResult();
+            m_queueProcessing = false;
+        }
+
+        public void AddMaintenanceTask(Func<TownKey, Task> startupTask)
+        {
+            m_startupTasks.Add(startupTask);
+        }
+
+        public async Task RunMaintenanceAsync()
```

Comment:
I am still confused about this.

If this method is ever called while the queue is in progress, then we'll end up with another queue running forever. There's nothing to prevent there from being 2 outstanding requests for processing queues, and they'll fight over the `m_queueProcessing` bool

Reply:
Yeah I didn’t make any changes yet because we were discussing it in the comments